databricks/koalas/frame.py

```diff
@@ -7819,15 +7815,20 @@ def _reindex_index(self, index):
 
         kser = ks.Series(list(index))
         labels = kser._internal.spark_frame.select(kser.spark.column.alias(index_column))
+        frame = self._internal.resolved_copy.spark_frame.drop(NATURAL_ORDER_COLUMN_NAME)
 
-        joined_df = self._internal.resolved_copy.spark_frame.drop(NATURAL_ORDER_COLUMN_NAME).join(
-            labels, on=index_column, how="right"
-        )
-        internal = self._internal.with_new_sdf(joined_df)
+        if fill_value is not None:
+            unmatched_labels = labels.subtract(frame.select(self._internal.spark_columns[0]))
+            filled = frame.join(unmatched_labels, on=index_column, how="right").fillna(fill_value)
+            existing = frame.join(labels, on=index_column, how="inner")
+            joined_df = existing.union(filled)
```

Comment:
I'm afraid there are too many joins here. How about:

```py
            frame_index_column = verify_temp_column_name(frame, "__frame_index_column__")
            frame = frame.withColumnRenamed(index_column, frame_index_column)

            temp_fill_value = verify_temp_column_name(frame, "__fill_value__")
            labels = labels.withColumn(temp_fill_value, F.lit(fill_value))

            frame_index_scol = scol_for(frame, frame_index_column)
            labels_index_scol = scol_for(labels, index_column)

            joined_df = frame.join(labels, on=[frame_index_scol == labels_index_scol], how="right")
            joined_df = joined_df.select(
                index_column,
                *[
                    F.when(
                        frame_index_scol.isNull() & labels_index_scol.isNotNull(),
                        scol_for(joined_df, temp_fill_value),
                    )
                    .otherwise(scol_for(joined_df, col))
                    .alias(col)
                    for col in self._internal.data_spark_column_names
                ]
            )
```

Reply:
Yes, nice way to do it ! It will be more efficient.
I update, thanks